File: shared/contracts/agent-protocol.ts
/**
 * Typed view of the agent <-> server WS protocol.
 *
 * This file is hand-maintained and the canonical TypeScript source.
 * It mirrors `shared/contracts/agent-protocol.json` exactly — the JSON
 * Schema is canonical wire-format documentation (used by ajv for runtime
 * validation and as input to C# codegen). When you edit the schema,
 * update this file and `shared/contracts/AgentProtocol.cs` in the same
 * commit. `npm run validate:contracts` checks consistency in CI.
 *
 * Versioned: bump the `Envelope.v` literal for breaking changes.
 */

export const PROTOCOL_VERSION = 1 as const;

export type MessageType =
  | 'hello'
  | 'welcome'
  | 'server_ping'
  | 'heartbeat'
  | 'assign'
  | 'ack'
  | 'progress'
  | 'log'
  | 'complete'
  | 'fail'
  | 'cancel'
  | 'pollLayers'
  | 'layers'
  | 'restart'
  | 'update'
  | 'startVisualisation'
  | 'cancelVisualisation'
  | 'visualisationReady'
  | 'visualisationFailed';

export type AgentRole = 'conversion' | 'layering' | 'receive' | 'visualiser';

export interface HelloData {
  machineId: string;
  nodeName: string;
  slots: number;
  formats: string[];
  roles: AgentRole[];
  agentVersion: string;
  rhinoVersion?: string;
}

export interface WelcomeData {
  sessionId: string;
  serverTime: string;
  heartbeatSeconds?: number;
}

export interface HeartbeatData {
  slotsBusy: number;
  cpuPct?: number;
  memUsedMb?: number;
}

export type JobKind = 'convert' | 'receive';

export interface AssignData {
  jobId: string;
  jobType?: JobKind;          // default 'convert' for backwards compat with Phase 3 agents
  slot: number;
  format: string;
  fileUrl?: string;           // not set for `receive`
  fileName?: string;
  orbitServerUrl: string;
  orbitToken: string;
  projectId: string;
  modelId: string;
  modelName?: string;
  receiveVersionId?: string;  // required when jobType === 'receive'
  outputFormats?: string[];   // ['3dm','step','ifc','glb']; for receive: primary output is outputFormats[0]
  outputUploadUrl?: string;   // POST endpoint the agent uses to deliver non-ORBIT outputs back to the server
  options?: {
    swapYZ?: boolean;
    quality?: 'sensible' | 'extreme';
    includedLayers?: string[];
    includeLayerDescendants?: boolean;
  };
}

export interface AckData {
  jobId: string;
  accepted: boolean;
  reason?: string;
}

export interface ProgressData {
  jobId: string;
  stage: string;
  percent?: number;
  message?: string;
}

export interface LogData {
  jobId: string;
  level: 'debug' | 'info' | 'warn' | 'error';
  message: string;
}

export interface CompleteData {
  jobId: string;
  versionUrl?: string;                 // primary ORBIT URL; omitted for pure receive jobs
  rootObjectId?: string;
  versionId?: string;
  outputs?: Record<string, string>;    // additional output file URLs by format ('3dm','glb',...)
  stats?: {
    objects?: number;
    blobs?: number;
    uploadBytes?: number;
    elapsedMs?: number;
  };
}

export interface FailData {
  jobId: string;
  error: string;
  stack?: string;
  retryable?: boolean;
}

export interface CancelData {
  jobId: string;
  reason?: string;
}

export interface PollLayersData {
  jobId: string;
  fileUrl: string;
  format: string;
}

/**
 * Server -> agent: cleanly exit the agent process. The Windows Scheduled
 * Task's `RestartCount=3` / `RestartInterval=1m` auto-respawns it; the
 * agent itself also schedules a small helper script that relaunches the
 * EXE after the process exits, so respawn is robust regardless of how
 * the task scheduler is configured. Reason is optional and surfaced in
 * agent logs.
 */
export interface RestartData {
  reason?: string;
}

/**
 * Server -> agent: check for a new release on GitHub Releases and apply
 * it if one is available. Reuses `Updater.CheckForUpdateAsync` +
 * `DownloadAndInstallAsync` — the same code path as the tray's
 * "Check for updates" menu item. `tag` is optional; when omitted the
 * agent picks the latest release.
 */
export interface UpdateData {
  tag?: string;
}

export interface LayerNode {
  name: string;
  fullPath?: string;
  color?: string;
  visible?: boolean;
  children?: LayerNode[];
}

export interface LayersData {
  jobId: string;
  layers: LayerNode[];
}

/* -------------------------------------------------------------------------- */
/* Visualiser (Phase A scaffold — orchestrator lands in Phase F/G)            */
/* -------------------------------------------------------------------------- */

/**
 * Server -> agent: spin up a Pixel Streaming session for an ORBIT version.
 * The agent imports the model into an Unreal template build, starts the
 * stream, and asynchronously replies with `visualisationReady` (or
 * `visualisationFailed`).
 */
export interface StartVisualisationData {
  runId: string;
  slot: number;
  orbitServerUrl: string;
  orbitToken: string;
  projectId: string;
  modelId: string;
  /** Optional ORBIT version to materialise. When omitted the agent picks the latest. */
  versionId?: string;
  /** UE template tag the orchestrator should run against (e.g. `v1.0.0-ue5.7`). */
  templateTag?: string;
  /** Public signalling URL the SPA will connect to; agent may echo it back when ready. */
  signallingUrl?: string;
  /** Max session lifetime in seconds; the orchestrator enforces hard tear-down at TTL. */
  ttlSeconds?: number;
}

/** Server -> agent: tear down a previously-started visualisation run. */
export interface CancelVisualisationData {
  runId: string;
  reason?: string;
}

/** Agent -> server: orchestrator is live; signalling URL is reachable. */
export interface VisualisationReadyData {
  runId: string;
  signallingUrl: string;
  streamerId?: string;
  expiresAt?: string;
}

/** Agent -> server: terminal failure during import / boot / streaming. */
export interface VisualisationFailedData {
  runId: string;
  error: string;
  stack?: string;
}

/* -------------------------------------------------------------------------- */
/* Discriminated envelope union                                                */
/* -------------------------------------------------------------------------- */

interface Base<T extends MessageType, D> {
  v: typeof PROTOCOL_VERSION;
  type: T;
  id?: string;
  ts?: string;
  data: D;
}

export type HelloMsg      = Base<'hello',      HelloData>;
export type WelcomeMsg    = Base<'welcome',    WelcomeData>;
export type HeartbeatMsg  = Base<'heartbeat',  HeartbeatData>;
export type AssignMsg     = Base<'assign',     AssignData>;
export type AckMsg        = Base<'ack',        AckData>;
export type ProgressMsg   = Base<'progress',   ProgressData>;
export type LogMsg        = Base<'log',        LogData>;
export type CompleteMsg   = Base<'complete',   CompleteData>;
export type FailMsg       = Base<'fail',       FailData>;
export type CancelMsg     = Base<'cancel',     CancelData>;
export type PollLayersMsg = Base<'pollLayers', PollLayersData>;
export type LayersMsg     = Base<'layers',     LayersData>;
export type RestartMsg    = Base<'restart',    RestartData>;
export type UpdateMsg     = Base<'update',     UpdateData>;
export type StartVisualisationMsg  = Base<'startVisualisation',  StartVisualisationData>;
export type CancelVisualisationMsg = Base<'cancelVisualisation', CancelVisualisationData>;
export type VisualisationReadyMsg  = Base<'visualisationReady',  VisualisationReadyData>;
export type VisualisationFailedMsg = Base<'visualisationFailed', VisualisationFailedData>;

export type AgentToServerMsg =
  | HelloMsg | HeartbeatMsg | AckMsg | ProgressMsg | LogMsg | CompleteMsg | FailMsg | LayersMsg
  | VisualisationReadyMsg | VisualisationFailedMsg;

export type ServerToAgentMsg =
  | WelcomeMsg | AssignMsg | CancelMsg | PollLayersMsg | RestartMsg | UpdateMsg
  | StartVisualisationMsg | CancelVisualisationMsg;

export type AnyMsg = AgentToServerMsg | ServerToAgentMsg;

/* -------------------------------------------------------------------------- */
/* Helpers                                                                     */
/* -------------------------------------------------------------------------- */

export function envelope<T extends MessageType, D>(type: T, data: D, id?: string): Base<T, D> {
  return { v: PROTOCOL_VERSION, type, id, ts: new Date().toISOString(), data };
}

File: shared/contracts/AgentProtocol.cs
// SPDX-FileCopyrightText: REBUS-ORBIT
// Hand-maintained C# mirror of shared/contracts/agent-protocol.json + .ts.
// Edit all three in the same commit and let CI's `npm run validate:contracts`
// catch any drift in the MessageType union.

using System.Runtime.Serialization;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using Newtonsoft.Json.Serialization;

namespace PRISM.Contracts;

public static class Protocol
{
    public const int Version = 1;
}

[JsonConverter(typeof(StringEnumConverter), typeof(CamelCaseNamingStrategy))]
public enum MessageType
{
    Hello,
    Welcome,
    [EnumMember(Value = "server_ping")]
    ServerPing,
    Heartbeat,
    Assign,
    Ack,
    Progress,
    Log,
    Complete,
    Fail,
    Cancel,
    PollLayers,
    Layers,
    Restart,
    Update,
    // Visualiser (Phase A scaffold — handlers stub-ack `accepted: false`).
    // Phase G adds VisualisationEnded + SignallingFrame for the WS proxy.
    StartVisualisation,
    CancelVisualisation,
    VisualisationReady,
    VisualisationFailed,
    VisualisationEnded,
    SignallingFrame,
}

[JsonConverter(typeof(StringEnumConverter), typeof(CamelCaseNamingStrategy))]
public enum AgentRole
{
    Conversion,
    Layering,
    Receive,
    Visualiser,
}

[JsonConverter(typeof(StringEnumConverter), typeof(CamelCaseNamingStrategy))]
public enum LogLevel
{
    Debug,
    Info,
    Warn,
    Error,
}

public sealed class Envelope<TData>
{
    [JsonProperty("v")]    public int Version { get; set; } = Protocol.Version;
    [JsonProperty("type")] public MessageType Type { get; set; }
    [JsonProperty("id", NullValueHandling = NullValueHandling.Ignore)]
    public string? Id { get; set; }
    [JsonProperty("ts", NullValueHandling = NullValueHandling.Ignore)]
    public string? Timestamp { get; set; }
    [JsonProperty("data")] public TData Data { get; set; } = default!;

    public static Envelope<TData> New(MessageType type, TData data, string? id = null) => new()
    {
        Version = Protocol.Version,
        Type = type,
        Id = id,
        Timestamp = DateTime.UtcNow.ToString("o"),
        Data = data,
    };
}

/* -------------------------------------------------------------------------- */
/* Agent -> server                                                            */
/* -------------------------------------------------------------------------- */

public sealed class HelloData
{
    [JsonProperty("machineId")]    public string MachineId    { get; set; } = "";
    [JsonProperty("nodeName")]     public string NodeName     { get; set; } = "";
    [JsonProperty("slots")]        public int    Slots        { get; set; }
    [JsonProperty("formats")]      public string[] Formats    { get; set; } = Array.Empty<string>();
    [JsonProperty("roles")]        public AgentRole[] Roles   { get; set; } = Array.Empty<AgentRole>();
    [JsonProperty("agentVersion")] public string AgentVersion { get; set; } = "";
    [JsonProperty("rhinoVersion", NullValueHandling = NullValueHandling.Ignore)]
    public string? RhinoVersion { get; set; }
}

public sealed class HeartbeatData
{
    [JsonProperty("slotsBusy")] public int  SlotsBusy { get; set; }
    [JsonProperty("cpuPct",    NullValueHandling = NullValueHandling.Ignore)] public double? CpuPct    { get; set; }
    [JsonProperty("memUsedMb", NullValueHandling = NullValueHandling.Ignore)] public double? MemUsedMb { get; set; }
}

public sealed class AckData
{
    [JsonProperty("jobId")]    public string JobId    { get; set; } = "";
    [JsonProperty("accepted")] public bool   Accepted { get; set; }
    [JsonProperty("reason", NullValueHandling = NullValueHandling.Ignore)]
    public string? Reason { get; set; }
}

public sealed class ProgressData
{
    [JsonProperty("jobId")]   public string  JobId   { get; set; } = "";
    [JsonProperty("stage")]   public string  Stage   { get; set; } = "";
    [JsonProperty("percent", NullValueHandling = NullValueHandling.Ignore)] public double? Percent { get; set; }
    [JsonProperty("message", NullValueHandling = NullValueHandling.Ignore)] public string? Message { get; set; }
}

public sealed class LogData
{
    [JsonProperty("jobId")]   public string   JobId   { get; set; } = "";
    [JsonProperty("level")]   public LogLevel Level   { get; set; }
    [JsonProperty("message")] public string   Message { get; set; } = "";
}

public sealed class CompleteData
{
    [JsonProperty("jobId")]      public string JobId      { get; set; } = "";
    [JsonProperty("versionUrl",   NullValueHandling = NullValueHandling.Ignore)] public string? VersionUrl   { get; set; }
    [JsonProperty("rootObjectId", NullValueHandling = NullValueHandling.Ignore)] public string? RootObjectId { get; set; }
    [JsonProperty("versionId",    NullValueHandling = NullValueHandling.Ignore)] public string? VersionId    { get; set; }
    [JsonProperty("outputs",      NullValueHandling = NullValueHandling.Ignore)] public Dictionary<string, string>? Outputs { get; set; }
    [JsonProperty("stats",        NullValueHandling = NullValueHandling.Ignore)] public CompleteStats? Stats { get; set; }
}

public sealed class CompleteStats
{
    [JsonProperty("objects",     NullValueHandling = NullValueHandling.Ignore)] public int?  Objects     { get; set; }
    [JsonProperty("blobs",       NullValueHandling = NullValueHandling.Ignore)] public int?  Blobs       { get; set; }
    [JsonProperty("uploadBytes", NullValueHandling = NullValueHandling.Ignore)] public long? UploadBytes { get; set; }
    [JsonProperty("elapsedMs",   NullValueHandling = NullValueHandling.Ignore)] public long? ElapsedMs   { get; set; }
}

public sealed class FailData
{
    [JsonProperty("jobId")] public string JobId { get; set; } = "";
    [JsonProperty("error")] public string Error { get; set; } = "";
    [JsonProperty("stack",     NullValueHandling = NullValueHandling.Ignore)] public string? Stack     { get; set; }
    [JsonProperty("retryable", NullValueHandling = NullValueHandling.Ignore)] public bool?   Retryable { get; set; }
}

public sealed class LayerNode
{
    [JsonProperty("name")] public string Name { get; set; } = "";
    [JsonProperty("fullPath", NullValueHandling = NullValueHandling.Ignore)] public string?     FullPath { get; set; }
    [JsonProperty("color",    NullValueHandling = NullValueHandling.Ignore)] public string?     Color    { get; set; }
    [JsonProperty("visible",  NullValueHandling = NullValueHandling.Ignore)] public bool?       Visible  { get; set; }
    [JsonProperty("children", NullValueHandling = NullValueHandling.Ignore)] public LayerNode[]? Children { get; set; }
}

public sealed class LayersData
{
    [JsonProperty("jobId")]  public string      JobId  { get; set; } = "";
    [JsonProperty("layers")] public LayerNode[] Layers { get; set; } = Array.Empty<LayerNode>();
}

/* -------------------------------------------------------------------------- */
/* Server -> agent                                                            */
/* -------------------------------------------------------------------------- */

public sealed class WelcomeData
{
    [JsonProperty("sessionId")]  public string SessionId  { get; set; } = "";
    [JsonProperty("serverTime")] public string ServerTime { get; set; } = "";
    [JsonProperty("heartbeatSeconds", NullValueHandling = NullValueHandling.Ignore)]
    public int? HeartbeatSeconds { get; set; }
}

public sealed class AssignOptions
{
    [JsonProperty("swapYZ",                  NullValueHandling = NullValueHandling.Ignore)] public bool?     SwapYZ                  { get; set; }
    [JsonProperty("quality",                 NullValueHandling = NullValueHandling.Ignore)] public string?   Quality                 { get; set; }
    [JsonProperty("includedLayers",          NullValueHandling = NullValueHandling.Ignore)] public string[]? IncludedLayers          { get; set; }
    [JsonProperty("includeLayerDescendants", NullValueHandling = NullValueHandling.Ignore)] public bool?     IncludeLayerDescendants { get; set; }
}

public sealed class AssignData
{
    [JsonProperty("jobId")]          public string JobId          { get; set; } = "";
    [JsonProperty("jobType", NullValueHandling = NullValueHandling.Ignore)] public string? JobType { get; set; }
    [JsonProperty("slot")]           public int    Slot           { get; set; }
    [JsonProperty("format")]         public string Format         { get; set; } = "";
    [JsonProperty("fileUrl",  NullValueHandling = NullValueHandling.Ignore)] public string? FileUrl  { get; set; }
    [JsonProperty("fileName", NullValueHandling = NullValueHandling.Ignore)] public string? FileName { get; set; }
    [JsonProperty("orbitServerUrl")] public string OrbitServerUrl { get; set; } = "";
    [JsonProperty("orbitToken")]     public string OrbitToken     { get; set; } = "";
    [JsonProperty("projectId")]      public string ProjectId      { get; set; } = "";
    [JsonProperty("modelId")]        public string ModelId        { get; set; } = "";
    [JsonProperty("modelName",         NullValueHandling = NullValueHandling.Ignore)] public string?   ModelName        { get; set; }
    [JsonProperty("receiveVersionId",  NullValueHandling = NullValueHandling.Ignore)] public string?   ReceiveVersionId { get; set; }
    [JsonProperty("outputFormats",     NullValueHandling = NullValueHandling.Ignore)] public string[]? OutputFormats    { get; set; }
    [JsonProperty("outputUploadUrl",   NullValueHandling = NullValueHandling.Ignore)] public string?   OutputUploadUrl  { get; set; }
    [JsonProperty("options",           NullValueHandling = NullValueHandling.Ignore)] public AssignOptions? Options     { get; set; }
}

public sealed class CancelData
{
    [JsonProperty("jobId")] public string JobId { get; set; } = "";
    [JsonProperty("reason", NullValueHandling = NullValueHandling.Ignore)]
    public string? Reason { get; set; }
}

public sealed class PollLayersData
{
    [JsonProperty("jobId")]   public string JobId   { get; set; } = "";
    [JsonProperty("fileUrl")] public string FileUrl { get; set; } = "";
    [JsonProperty("format")]  public string Format  { get; set; } = "";
}

/// <summary>
/// Server -> agent: cleanly exit the agent process. The Windows
/// Scheduled Task respawns it within ~1 min, and the agent also
/// schedules a small helper script that relaunches the EXE after
/// the process exits.
/// </summary>
public sealed class RestartData
{
    [JsonProperty("reason", NullValueHandling = NullValueHandling.Ignore)]
    public string? Reason { get; set; }
}

/// <summary>
/// Server -> agent: check GitHub Releases for a new agent build and
/// apply it if one is available. Reuses Updater.CheckForUpdateAsync +
/// DownloadAndInstallAsync (the same code path as the tray menu's
/// "Check for updates"). <see cref="Tag"/> optionally pins a specific
/// release tag; when null/empty the agent picks the latest.
/// </summary>
public sealed class UpdateData
{
    [JsonProperty("tag", NullValueHandling = NullValueHandling.Ignore)]
    public string? Tag { get; set; }
}

/* -------------------------------------------------------------------------- */
/* Visualiser (Phase A scaffold)                                              */
/*                                                                            */
/* These envelopes describe the future signalling between PRISM server and    */
/* a Visualiser agent that hosts the Unreal Engine + Pixel Streaming          */
/* orchestrator. In Phase A the agent acks with `accepted: false` until the   */
/* orchestrator binary lands in Phase F/G.                                    */
/* -------------------------------------------------------------------------- */

/// <summary>
/// Server -> agent: spin up a Pixel Streaming session for an ORBIT version.
/// The agent imports the model into an Unreal template build, starts the
/// stream, and replies (asynchronously) with <see cref="VisualisationReadyData"/>
/// when the signalling URL is reachable.
/// </summary>
public sealed class StartVisualisationData
{
    [JsonProperty("runId")]          public string RunId          { get; set; } = "";
    [JsonProperty("slot")]           public int    Slot           { get; set; }
    [JsonProperty("orbitServerUrl")] public string OrbitServerUrl { get; set; } = "";
    [JsonProperty("orbitToken")]     public string OrbitToken     { get; set; } = "";
    [JsonProperty("projectId")]      public string ProjectId      { get; set; } = "";
    [JsonProperty("modelId")]        public string ModelId        { get; set; } = "";
    [JsonProperty("versionId",       NullValueHandling = NullValueHandling.Ignore)] public string? VersionId      { get; set; }
    [JsonProperty("templateTag",     NullValueHandling = NullValueHandling.Ignore)] public string? TemplateTag    { get; set; }
    [JsonProperty("signallingUrl",   NullValueHandling = NullValueHandling.Ignore)] public string? SignallingUrl  { get; set; }
    [JsonProperty("ttlSeconds",      NullValueHandling = NullValueHandling.Ignore)] public int?    TtlSeconds     { get; set; }
}

/// <summary>
/// Server -> agent: tear down a previously-started visualisation run.
/// </summary>
public sealed class CancelVisualisationData
{
    [JsonProperty("runId")] public string RunId { get; set; } = "";
    [JsonProperty("reason", NullValueHandling = NullValueHandling.Ignore)]
    public string? Reason { get; set; }
}

/// <summary>
/// Agent -> server: the orchestrator has imported the model, started UE,
/// and the signalling endpoint is live. Carries the URL the client SPA
/// should connect its WebRTC negotiation to.
/// </summary>
public sealed class VisualisationReadyData
{
    [JsonProperty("runId")]         public string RunId         { get; set; } = "";
    [JsonProperty("signallingUrl")] public string SignallingUrl { get; set; } = "";
    [JsonProperty("streamerId", NullValueHandling = NullValueHandling.Ignore)] public string? StreamerId { get; set; }
    [JsonProperty("expiresAt",  NullValueHandling = NullValueHandling.Ignore)] public string? ExpiresAt  { get; set; }
}

/// <summary>
/// Agent -> server: the orchestrator could not start (import failed,
/// no GPU, UE crashed, etc.). Terminal state for the run.
/// </summary>
public sealed class VisualisationFailedData
{
    [JsonProperty("runId")] public string RunId { get; set; } = "";
    [JsonProperty("error")] public string Error { get; set; } = "";
    [JsonProperty("stack", NullValueHandling = NullValueHandling.Ignore)] public string? Stack { get; set; }
}

/// <summary>
/// Agent -> server: a previously-streaming run ended cleanly (TTL expired,
/// UE exited, browser disconnected, admin cancel). Terminal state for the run.
/// </summary>
public sealed class VisualisationEndedData
{
    [JsonProperty("runId")] public string RunId { get; set; } = "";
    [JsonProperty("reason", NullValueHandling = NullValueHandling.Ignore)]
    public string? Reason { get; set; }
}

/// <summary>
/// Bidirectional opaque WebRTC signalling envelope. PRISM does not parse
/// the Pixel Streaming sub-protocol — the server wraps browser frames
/// into one of these, the agent unwraps and forwards to the local Cirrus
/// WS, and the reverse direction does the same. Exactly one of
/// <see cref="Payload"/> (text) / <see cref="PayloadB64"/> (binary) is
/// set per frame.
/// </summary>
public sealed class SignallingFrameData
{
    [JsonProperty("runId")] public string RunId { get; set; } = "";
    [JsonProperty("payload",    NullValueHandling = NullValueHandling.Ignore)] public string? Payload    { get; set; }
    [JsonProperty("payloadB64", NullValueHandling = NullValueHandling.Ignore)] public string? PayloadB64 { get; set; }
}
